### PR TITLE
add libdfl

### PR DESCRIPTION
**Pull request by r-rojo**
Add first iteration of libdfl

### CODE REVIEW

File: libdfl/opae_dfl.c [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+	if (attr_str) {
+		char *endptr = NULL;
+		*value = strtoul(attr_str, &endptr, 0);
+		return endptr == attr_str ? FPGA_EXCEPTION : FPGA_OK;
+	}
+	return FPGA_NOT_FOUND;
+}
+
+
+STATIC int dfl_device_parse_id(dfl_device *device)
```
Comment (tswhison) [inline libdfl/opae_dfl.c:205]: This returns fpga_result, but the return value is typed as int.

File: libdfl/opae_dfl.c [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+		OPAE_ERR("Error parsing vendor_id for device: %s",
+			 udev_device_get_syspath(device->dev));
+		return FPGA_EXCEPTION;
+	}
+	device->vendor_id = value;
+
+	return FPGA_OK;
+}
+
+static int dfl_get_errors(dfl_device *dev)
```
Comment (tswhison) [inline libdfl/opae_dfl.c:228]: STATIC so that it can be tested.

File: libdfl/opae_dfl.c [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+				err->next = NULL;
+				if (dev->errors) {
+					dfl_error *tmp = dev->errors;
+					while(tmp->next) tmp = tmp->next;
+					tmp->next = err;
+				} else {
+					dev->errors = err;
+				}
+			} else {
+				printf("error allocating memory for error struct\n");
```
Comment (tswhison) [inline libdfl/opae_dfl.c:252]: OPAE_ERR and exit with error code.

File: libdfl/opae_dfl.c [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+		const char *attr = udev_list_entry_get_name(le);
+		if (attr && !fnmatch(DFL_ERRORS_PATTERN, attr, FNM_EXTMATCH)) {
+			dfl_error *err = (dfl_error*)malloc(sizeof(dfl_error));
+			if (err) {
+				++dev->num_errors;
+				err->attr = attr;
+				err->next = NULL;
+				if (dev->errors) {
+					dfl_error *tmp = dev->errors;
+					while(tmp->next) tmp = tmp->next;
```
Comment (tswhison) [inline libdfl/opae_dfl.c:246]: Save a pointer to the last list entry so that you don't have to walk the list each time.

File: libdfl/opae_dfl.c [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+}
+
+dfl_device *dfl_device_add(dfl_device *e, const char *path)
+{
+	dfl_device *dfl = dfl_device_new(path);
+	if (!dfl) {
+		return NULL;
+	}
+	if (e) {
+		while(e->next) {
```
Comment (tswhison) [inline libdfl/opae_dfl.c:338]: Is it strictly necessary to insert at the end of the list?  If not, why not insert at the beginning to save time?

File: libdfl/opae_dfl.c [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+	if (e->type == FPGA_DEVICE) return NULL;
+	const char *region = udev_device_get_syspath(e->region);
+	DIR *region_dir = opendir(region);
+	struct dirent *de = NULL;
+	if (!region_dir) {
+		OPAE_ERR("Could not open '%s'\n", region);
+		return NULL;
+	}
+	char fme_path[PATH_MAX] = {0};
+	while ((de = readdir(region_dir))) {
```
Comment (tswhison) [inline libdfl/opae_dfl.c:429]: Static analysis may complain about this sort of thing.

File: libdfl/opae_dfl.c [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+{
+	int i = 0;
+	for (const char *ptr = guid_str; i < 16; ++i, ptr += 2)
+		if (!sscanf(ptr, "%02hhx", (uint8_t *)&guid[i])) {
+			return 0;
+		}
+	return i;
+}
+
+int dfl_device_get_compat_id(dfl_device *d, fpga_guid guid)
```
Comment (tswhison) [inline libdfl/opae_dfl.c:451]: Return value is int, but you're returning fpga_result.

File: libdfl/opae_dfl.c [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+		}
+
+	}
+out_free:
+	if (dev) udev_device_unref(dev);
+	udev_enumerate_unref(ue);
+	return res;
+}
+
+int dfl_device_get_afu_id(dfl_device *d, fpga_guid guid)
```
Comment (tswhison) [inline libdfl/opae_dfl.c:499]: int, but returning fpga_result.

File: tests/libudev-mock/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+## IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+## ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+## LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+## CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+## SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+## INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+## CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+## POSSIBILITY OF SUCH DAMAGE.
+add_library(udev SHARED
```
Comment (tswhison) [inline tests/libudev-mock/CMakeLists.txt:26]: Not using the opae macros?

File: tests/libdfl/test_libdfl.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
+// Copyright(c) 2021, Intel Corporation
```
Comment (tswhison) [inline tests/libdfl/test_libdfl.cpp:1]: What's the current coverage number?

Comment (r-rojo) [reply tests/libdfl/test_libdfl.cpp:1]: 73%

File: libdfl/opae_dfl.c (hunk 6 above)
Comment (r-rojo) [reply libdfl/opae_dfl.c:429]: I'll wait to see if/when it does

File: libdfl/opae_dfl.c (hunk 5 above)
Comment (r-rojo) [reply libdfl/opae_dfl.c:338]: not strictly necessary... I just wanted to keep it in the order it was added (with last added at the tail).
The time savings are negligible

File: libdfl/opae_dfl.c (hunk 4 above)
Comment (r-rojo) [reply libdfl/opae_dfl.c:246]: I don't mind walking it

File: libdfl/opae_dfl.c [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+	return FPGA_NOT_FOUND;
+}
+
+STATIC int udev_parse_attr32(struct udev_device *dev, const char *attr, uint32_t *value)
+{
+	const char *attr_str = udev_device_get_sysattr_value(dev, attr);
+	if (attr_str) {
+		char *endptr = NULL;
+		*value = strtoul(attr_str, &endptr, 0);
+		return endptr == attr_str ? FPGA_EXCEPTION : FPGA_OK;
```
[the remark this answers is not in the document]
Comment (r-rojo) [reply libdfl/opae_dfl.c:199]: from the man page:
```
If there were no digits at all, strtoul() stores the original value of nptr in *endptr (and returns 0)
```

File: libdfl/opae_dfl.c [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+//}
+
+static int udev_parse_attr64(struct udev_device *dev, const char *attr, uint64_t *value)
+{
+	const char *attr_str = udev_device_get_sysattr_value(dev, attr);
+	if (attr_str) {
+		if (strlen(attr_str)) {
+			char *endptr = NULL;
+			*value = strtoull(attr_str, &endptr, 0);
+			return endptr == attr_str ? FPGA_EXCEPTION : FPGA_OK;
```
[the remark this answers is not in the document]
Comment (r-rojo) [reply libdfl/opae_dfl.c:186]: from the man page:
```
If there were no digits at all, strtoul() stores the original value of nptr in *endptr (and returns 0)
```

File: libdfl/opae_dfl.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (tswhison) [reply libdfl/opae_dfl.c:199]: My comment still stands.  If it didn't scan anything (the scan string was invalid), then this is still an error, no?

File: libdfl/opae_dfl.c [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+
+STATIC int udev_direct_read_attr64(struct udev_device *dev,
+				   const char *attr, uint64_t *value)
+{
+	int res = FPGA_EXCEPTION;
+	char *buffer = NULL;
+	if (udev_direct_read_attr(dev, attr, &buffer)) {
+		char *endptr = NULL;
+		*value = strtoull(buffer, &endptr, 0);
+		if (endptr > buffer)
```
[the remark this answers is not in the document]
Comment (r-rojo) [reply libdfl/opae_dfl.c:157]: Or something like `12\n`

File: libdfl/opae_dfl.c [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+	}
+	res = FPGA_OK;
+
+out:
+	regfree(&re);
+	return res;
+}
+
+
+void udev_print_list(struct udev_list_entry *le)
```
[the remark this answers is not in the document]
Comment (r-rojo) [reply libdfl/opae_dfl.c:108]: This will probably go away later. Just using it for debug convenience. I'll add a TODO.

**Review by tswhison — APPROVED** [review]
👍

**Pull request merged**